### PR TITLE
check swap path

### LINKED ISSUE #1875: Require input token and output token are different
For DEX, the input token and final output token should be different.

### PR DESCRIPTION
**Pull request by xlc**
closes #1875

### CODE REVIEW

File: modules/dex/src/lib.rs [hunk 1]
```diff
@@ -1176,6 +1174,17 @@ impl<T: Config> Pallet<T> {
 		Ok(supply_amounts)
 	}
 
+	fn validate_path(path: &[CurrencyId]) -> DispatchResult {
+		let path_length = path.len();
+		ensure!(
+			path_length >= 2 && path_length <= T::TradingPathLimit::get().saturated_into(),
+			Error::<T>::InvalidTradingPathLength
+		);
+		ensure!(path[0] != path[path_length - 1], Error::<T>::InvalidTradingPath);
```
Comment (zjb0807) [inline modules/dex/src/lib.rs:1183]:
```suggestion
		ensure!(path.iter().unique().len() == path_length, Error::<T>::InvalidTradingPath);
```

and make it `public` for the `validate_path`

Comment (shaunxw) [reply modules/dex/src/lib.rs:1183]: Checking unique is much worse than getting elements & compare in performance.

Comment (zjb0807) [reply modules/dex/src/lib.rs:1183]: I think `kUSD, KAR, KAR, KSM` and `KUSD, KSM, KUSD, KAR` should be checked.

Comment (shaunxw) [inline modules/dex/src/lib.rs:1183]:
```suggestion
		ensure!(path.get(0) != path.get(path_length - 1), Error::<T>::InvalidTradingPath);
```

`vector[n]` could panic, I think we should use `get` instead

Comment (shaunxw) [reply modules/dex/src/lib.rs:1183]: It's common to trade across two pools. I don't see the benefits of disabling them.

Comment (xlc) [reply modules/dex/src/lib.rs:1183]: we checked length >= 2 so it shouldn't panic, but yeah can still improve it

Comment (zjb0807) [reply modules/dex/src/lib.rs:1183]: `KUSD, KSM, KSM, KAR` will generate the tradingpair `KUSD, KSM`, `KSM, KSM`,  `KSM,KAR`. But the `KSM, KSM` is invalid.
`KUSD, KSM, KUSD, KAR` should be `KUSD, KAR`.

**Review by zjb0807 — APPROVED** [review]

**Review by shaunxw — APPROVED** [review]

**Pull request merged**